onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -1159,7 +1166,10 @@ std::unique_lock<OrtMutex> TensorrtExecutionProvider::GetEngineBuildLock() const
 
 common::Status TensorrtExecutionProvider::Compile(const std::vector<Node*>& fused_nodes,
                                                   std::vector<NodeComputeInfo>& node_compute_funcs) {
-  for (const auto* fused_node : fused_nodes) {
+  int fused_nodes_size = fused_nodes.size();
+  cuda_graphs_.reserve(fused_nodes_size);
+  for (int node_idx = 0; node_idx < fused_nodes_size; node_idx++) {
+    const auto* fused_node = fused_nodes[node_idx];
```

Comment:
In general, how does the TRT EP handle control flow nodes ? I fear we must explicitly not support using cuda graphs for models with control flow nodes as the graph captured for one input may not the same graph required for another input (because of the dynamic graph branching).

Reply:
agreed. we should explicitly exclude graphs with loops/conditionals.

Reply:
We may highlight the constrains for dynamic shape cases in document, so that users can choose to enable cuda graph or not.

Reply:
it would be nice to enforce constraints (doesn't support dynamic shapes and dynamic graphs) in code rather than punting to user/documentation. Let's see if there's a reasonable balance that can be achieved here.

Reply:
Here is a little bit tricky. Dynamic shape model is okay. Only when input shapes of incoming data change, cuda graph needs to be recaptured, so the check has to be done in runtime. So far I haven't seen any APIs that can check existing cuda graph's profile. There is API to update executable graph, but we can't afford to update graph for each enqueue.